### PR TITLE
cmake: multi-image: Fix CONFIG_APP_LINK_WITH_ mechanism

### PR DESCRIPTION
**Pull request by SebastianBoe**
Fix CONFIG_APP_LINK_WITH_ mechanism

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

### CODE REVIEW

File: cmake/app/boilerplate.cmake [hunk 1]
```diff
@@ -395,8 +395,6 @@ set_property(TARGET ${IMAGE}app PROPERTY ARCHIVE_OUTPUT_DIRECTORY ${IMAGE}app)
 
 add_subdirectory(${ZEPHYR_BASE} ${__build_dir})
 
-#TODO: Everything below here.
```
Comment (hakonfam) [inline cmake/app/boilerplate.cmake:398]: wow, wow, wow, so now everything is done? :smile:

File: cmake/app/boilerplate.cmake [hunk 2]
```diff
@@ -410,8 +408,25 @@ foreach(boilerplate_lib ${ZEPHYR_INTERFACE_LIBS_PROPERTY})
...
   # flags in the interface of a library.
-  #
-  string(TOUPPER ${boilerplate_lib} boilerplate_lib_upper_case) # Support lowercase lib names
+
+  # 'boilerplate_lib' is formatted as '0_mbedtls' (for instance). But
+  # the Kconfig options are formatted as
+  # 'CONFIG_APP_LINK_WITH_MBEDTLS'. So we need to strip the '0_' and
+  # convert to upper case.
+
+  # Match the non-image'ified library name.
```
Comment (hakonfam) [inline cmake/app/boilerplate.cmake]:
```suggestion
  # Match the non-image'ified library name. E.g. '1_mylib' -> 'mylib'
```

File: cmake/app/boilerplate.cmake [hunk 3]
```diff
@@ -410,8 +408,25 @@ foreach(boilerplate_lib ${ZEPHYR_INTERFACE_LIBS_PROPERTY})
...
+  string(REGEX MATCH
+	"^${IMAGE}(.*)"
+	unused_out_var
+	${boilerplate_lib}
+	)
+  if(CMAKE_MATCH_1)
+	set(boilerplate_lib_without_image ${CMAKE_MATCH_1})
+  else()
+	message(FATAL_ERROR "Internal error. Expected '${boilerplate_lib}' to match '^${IMAGE}(.*)'")
```
Comment (hakonfam) [inline cmake/app/boilerplate.cmake]: This begs to be outdated once the code changes, either use av variable for the regular expression, or remove it from the error message.

**Review by hakonfam — CHANGES_REQUESTED** [review]

File: cmake/app/boilerplate.cmake (hunk 1 above)
Comment (SebastianBoe) [reply cmake/app/boilerplate.cmake:398]: hehe

**Comment by SebastianBoe** [comment]
@hakonfam : All review comments have been addressed.

**Pull request merged**